### PR TITLE
expression: fix wrong result of floor

### LINKED ISSUE #6595: result of ceil(decimal) is not correct
## 1. What did you do?
```sql
drop table if exists t;
create table t(a decimal(40,20) UNSIGNED);
insert into t values(2.99999999900000000000);
select a, ceil(a) from t where ceil(a) > 1;
```
NOTE: both `a` and `ceil(a)` is *not correct*.

## 2. What did you expect to see?

```sql
MySQL(localhost:3306) > select a, ceil(a) from t where ceil(a) > 1;
+------------------------+---------+
| a                      | ceil(a) |
+------------------------+---------+
| 2.99999999900000000000 |       3 |
+------------------------+---------+
1 row in set (0.00 sec)

MySQL(localhost:3306) > select a, ceil(a) from t;
+------------------------+---------+
| a                      | ceil(a) |
+------------------------+---------+
| 2.99999999900000000000 |       3 |
+------------------------+---------+
1 row in set (0.00 sec)
```

## 3. What did you see instead?
```sql
TiDB(localhost:4000) > select a, ceil(a) from t where ceil(a) > 1;
+------------------------+---------+
| a                      | ceil(a) |
+------------------------+---------+
| 4.99999999900000000000 |       4 |
+------------------------+---------+
1 row in set (0.00 sec)

TiDB(localhost:4000) > select a, ceil(a) from t;
+------------------------+---------+
| a                      | ceil(a) |
+------------------------+---------+
| 3.99999999900000000000 |       3 |
+------------------------+---------+
1 row in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by zz-jason**
fix #6595 

```sql
drop table if exists t;
create table t(a decimal(40,20), b bigint);
insert into t values(-2.99999990000000000000, -1);
```

Before this PR:
```sql
TiDB(localhost:4000) > select floor(a), floor(a), floor(a) from t;
+----------+----------+----------+
| floor(a) | floor(a) | floor(a) |
+----------+----------+----------+
|       -3 |       -4 |       -5 |
+----------+----------+----------+
1 row in set (0.00 sec)

TiDB(localhost:4000) > select b, floor(b) from t;
+------+----------------------+
| b    | floor(b)             |
+------+----------------------+
|   -1 | 18446744073709551615 |
+------+----------------------+
1 row in set (0.00 sec)
```

After this PR:
```sql
TiDB(localhost:4000) > select floor(a), floor(a), floor(a) from t;
+----------+----------+----------+
| floor(a) | floor(a) | floor(a) |
+----------+----------+----------+
|       -3 |       -3 |       -3 |
+----------+----------+----------+
1 row in set (0.01 sec)

TiDB(localhost:4000) > select b, floor(b) from t;
+------+----------+
| b    | floor(b) |
+------+----------+
|   -1 |       -1 |
+------+----------+
1 row in set (0.00 sec)
```

### CODE REVIEW

**Comment by zz-jason** [comment]
@XuHuaiyu @winoros @tiancaiamao PTAL

File: expression/builtin_math.go [hunk 1]
```diff
@@ -702,7 +702,14 @@ func (b *builtinFloorIntToDecSig) Clone() builtinFunc {
 // See https://dev.mysql.com/doc/refman/5.7/en/mathematical-functions.html#function_floor
 func (b *builtinFloorIntToDecSig) evalDecimal(row types.Row) (*types.MyDecimal, bool, error) {
 	val, isNull, err := b.args[0].EvalInt(b.ctx, row)
-	return types.NewDecFromUint(uint64(val)), isNull, errors.Trace(err)
+	if isNull || err != nil {
+		return nil, true, errors.Trace(err)
+	}
+
+	if mysql.HasUnsignedFlag(b.args[0].GetType().Flag) || val >= 0 {
```
Comment (winoros) [inline expression/builtin_math.go]: Why we need to check `val > 0` when it has `unsigned` flag

Comment (zz-jason) [reply expression/builtin_math.go]: Oh.. `val >= 0` is checked only when `mysql.HasUnsignedFlag(b.args[0].GetType().Flag)` is false

**Pull request merged**